Sai Venkatesh- Backend - Improve speed and function of deleting multiple blue squares

SFA23SCM35V opened the pull request:
# Description
Blue squares used to come back, now they just delete really slowly or not at all

## Related PRS (if any):
This backend PR is related to the #[2774](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2774) frontend PR.

…

## Main changes explained:
Used async/await consistently, removing the callback approach from findById.
Simplified error handling by wrapping the entire block in a try/catch.
Replaced redundant return statements with direct res responses for permission and record checks.
Improved readability and structure by removing unnecessary comments and logs.
…

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as Owner user
5. Profile>Blue Square Section>Delete a bunch in a row
6. verify functionality 


https://www.loom.com/share/02a9fd342bf84425abdf547566c85035

nathanah commented on src/controllers/userProfileController.js:
```diff
@@ -1686,37 +1686,82 @@ const userProfileController = function (UserProfile, Project) {
-    }
-    const { userId, blueSquareId } = req.params;
-    // console.log(userId, blueSquareId);
+  // const deleteInfringements = async function (req, res) {
+  //   if (!(await hasPermission(req.body.requestor, 'deleteInfringements'))) {
+  //     res.status(403).send('You are not authorized to delete blue square');
+  //     return;
+  //   }
+  //   const { userId, blueSquareId } = req.params;
+  //   // console.log(userId, blueSquareId);
+
+  //   UserProfile.findById(userId, async (err, record) => {
+  //     if (err || !record) {
+  //       res.status(404).send('No valid records found');
+  //       return;
+  //     }
+
+  //     const originalinfringements = record?.infringements ?? [];
+
+  //     record.infringements = originalinfringements.filter(
+  //       (infringement) => !infringement._id.equals(blueSquareId),
+  //     );
+
+  //     record
+  //       .save()
+  //       .then((results) => {
+  //         userHelper.notifyInfringements(originalinfringements, results.infringements);
+  //         res.status(200).json({
+  //           _id: record._id,
+  //         });
+  //       })
+  //       .catch((error) => res.status(400).send(error));
+  //   });
+  // };
```
```suggestion
```
Deleting this old code instead of just commenting it out and duplicating it makes the commit history slightly easier to see/compare changes and reduces clutter from old code.

nathanah commented on src/controllers/userProfileController.js:
```diff
@@ -1686,37 +1686,82 @@ const userProfileController = function (UserProfile, Project) {
-        return;
-      }
+  
+/*
+Used async/await consistently, removing the callback approach from findById.
+Simplified error handling by wrapping the entire block in a try/catch.
+Replaced redundant return statements with direct res responses for permission and record checks.
+Improved readability and structure by removing unnecessary comments and logs.
 
-      const originalinfringements = record?.infringements ?? [];
+*/
```
It might be good to remove this comment. This type of info should be in the PR/commit descriptions because parts referring to removing code won't make sense outside the context of this PR.

nathanah reviewed (changes requested):
Just remove some of the comments to clean things up and I'll approve. The code refactor looks good.

Ankuriboh reviewed (dismissed):
Refactoring existing backend code, tested and should be ready to merge. Before that, please kindly remove all the unnecessary comments from the code.

Also, as mentioned in the corresponding frontend PR, please kindly branching out in this repo instead of a forked one. Thanks!

sheetalmangate commented:
In the code refactor, a try-catch block is used for all code within the function. what If the .save() method fails, it should be handled either in a specific try-catch block or by using a callback function.

SFA23SCM35V commented:
@nathanah 
1. Consistent use of async/await and removal of callbacks
Improved Readability and Debugging: By using async/await consistently instead of mixing callbacks with promises, the code becomes more linear and easier to understand. This makes it easier to spot bugs and reason about the program flow, which could reduce the time spent debugging.
Reduced Nested Callbacks (Callback Hell): Callbacks often lead to deeply nested code structures, which are harder to maintain and debug. Using async/await flattens the code, making it more readable and maintainable.
Impact on speed: While this change doesn't directly improve execution speed, it helps avoid bugs and potential inefficiencies in how asynchronous code executes, making your application less prone to errors like callback mismanagement (e.g., multiple responses).

2. Simplified Error Handling with try/catch
Less Complexity: Wrapping the whole asynchronous code block in a try/catch block allows you to handle errors in a single place instead of adding error handling to every callback. This reduces cognitive load for the developer and ensures that errors are caught and handled properly.
Consistency: Simplified error handling ensures that all errors (network issues, database errors, etc.) are caught uniformly, which helps prevent unexpected crashes that could slow down the app.
Impact on speed: While simplifying error handling itself doesn't directly speed up the code, better error management means that the app may handle failure cases (such as database errors) faster and more predictably, preventing performance degradation from unhandled exceptions.

3. Removed Redundant return Statements
Improved Flow: By directly sending responses using res.status(...).send(...) instead of unnecessary return statements, the code becomes more straightforward, reducing confusion.
Code Simplicity: The more concise code means fewer instructions for the JavaScript engine to interpret, and while this is a micro-optimization, cleaner code usually leads to fewer bugs and easier performance profiling.
Impact on speed: There's minimal direct speed improvement here, but a more straightforward control flow ensures that there are no unnecessary execution paths, which can help reduce response time.

4. Reduced Unnecessary Comments and Logs
Cleaner Code: Removing unnecessary comments and logs can speed up the development and testing process by reducing noise. During debugging, logging statements can slow down performance slightly, especially if they're placed in critical or frequent parts of the code, like inside loops or asynchronous functions.
Less Disk I/O or Console Overhead: While logging helps during development, leaving unnecessary logs in production can introduce overhead, especially if logging is synchronous or involves writing to disk or remote services.
Impact on speed: Removing unnecessary logging can marginally improve performance, especially if logging was blocking or performed during high-traffic operations. This is a small gain, but it can be noticeable in highly concurrent systems.


@nathanah

Consolidated Destructuring for Props
By destructuring props at the start of the function:
const { blueSquares, handleBlueSquare, hasPermission } = props;
Benefit: This reduces redundant lookups like props.hasPermission, making the code more concise and slightly faster at runtime (though the performance gain is negligible in modern JavaScript engines). The main benefit here is readability and easier debugging, which reduces development time when modifying or maintaining the delete functionality.

Simplified handleOnClick by Reducing Repeated Conditions
Before: You had multiple conditions handling edit, delete, and view actions separately.
After: You combined similar conditions into a single call, like so:
if (canEditInfringements || canDeleteInfringements) {
handleBlueSquare(true, 'modBlueSquare', blueSquare._id);
} else {
handleBlueSquare(true, 'viewBlueSquare', blueSquare._id);
}

Benefit: This makes the logic easier to follow, reduces unnecessary branches, and ensures quicker execution when deleting. Although the performance improvement is minimal, cleaner logic reduces the chance of bugs, making the delete functionality more reliable and easier to optimize in the future.

Moved Formatting Logic Inside JSX
Before: You were checking for conditions and formatting strings in separate blocks.
After: You moved this into the JSX for a more streamlined rendering process:
{blueSquare.createdDate ? ${formatCreatedDate(blueSquare.createdDate)}:  : ''}
{blueSquare.description}

Benefit: This minimizes unnecessary intermediate logic and lets React handle the rendering more efficiently. This doesn’t directly speed up the deletion but makes rendering the component more efficient, especially in complex UIs with lots of elements. By reducing unnecessary overhead in rendering, React can potentially free up resources faster when performing state updates like deletion.

Removed Redundant Conditions and Comments
Benefit: Removing unused or redundant conditions reduces the number of checks the system has to make when rendering or handling click events. While this has minimal direct performance impact, it improves the overall flow of the code, making it easier to read, debug, and optimize in the future.
How These Changes Affect Deletion Speed
Reduced Logical Overhead: Simplifying the conditions for clicking (i.e., when a user clicks to delete a blue square) means the event handler is more straightforward, with fewer branching conditions. This can slightly reduce the time it takes for the function to determine what action to perform (edit, delete, or view), making deletion a bit faster in terms of execution.

Better Maintainability: Cleaner code is easier to modify and optimize. If you ever need to further optimize the delete functionality (e.g., through better state management or using React.memo to prevent unnecessary re-renders), having a cleaner base will make these optimizations more effective and easier to implement.

If you still feel it doesn't improve the speed, feel free to suggest changes.

SFA23SCM35V commented:
@Ankuriboh 
removed the comments and will create a new branch in this repo from now
if everything works well please Approve the PR.

samarth9201 commented on src/controllers/userProfileController.js:
```diff
@@ -1686,37 +1686,42 @@ const userProfileController = function (UserProfile, Project) {
+      // Respond with success
+      res.status(200).json({ _id: updatedRecord._id });
+    } catch (error) {
+      res.status(400).send(error.message);
```
This could possibly leak some sensitive information. For example, it can return stack traces which can contain information about database and environment variables. Instead send a standard error message.

samarth9201 reviewed (commented):
The code works as expected, just a small suggestion from security point of view. But overall looks okay.

![Screenshot (86)](https://github.com/user-attachments/assets/ad35e537-0abc-4fcc-8a29-2f58fea58eca)

one-community commented:
Thank you all, merging!
